src/core/Tree/Tree.tsx

```diff
@@ -0,0 +1,216 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps, useTheme } from '../utils';
+import '@itwin/itwinui-css/css/tree.css';
+import cx from 'classnames';
+
+export const TreeContext = React.createContext<
+  | {
+      nodeDepth: number;
+      subNodeIds?: string[];
+      hasSubNodes: boolean;
+    }
+  | undefined
+>(undefined);
+
+export type NodeData<T> = {
+  subNodes?: Array<T>;
+  nodeId: string;
+  node: T;
+  isExpanded?: boolean;
+  isDisabled?: boolean;
+  isSelected?: boolean;
+  hasSubNodes: boolean;
+};
+
+type FlatNode<T> = {
+  nodeData: NodeData<T>;
+  depth: number;
+  subNodeIds?: Array<string>;
+};
+
+export type TreeProps<T> = {
+  /**
+   * Custom renderer for tree nodes, recommended to use TreeNode component.
+   */
+  nodeRenderer: (props: NodeData<T>) => JSX.Element;
+  /**
+   * Items inside tree.
+   */
+  data: T[];
+  /**
+   * Get the NodeData.
+   */
+  getNode: (node: T) => NodeData<T>;
+} & CommonProps;
+
+/**
+ * Tree
+ * @example
+  <Tree<TreeData>
+    data={data}
+    getNode={getNode}
+    nodeRenderer={(props) => (
+      <TreeNode
+        nodeId={props.nodeId}
+        label={props.node.label}
+        sublabel={props.node.subLabel}
+        onNodeExpanded={onNodeExpanded}
+        onNodeSelected={onSelectedNodeChange}
+        isDisabled={props.isDisabled}
+        isExpanded={props.isExpanded}
+        isSelected={props.isSelected}
+        nodeCheckbox={
+          <Checkbox variant='eyeball' disabled={props.isDisabled} />
+        }
+        icon={<SvgPlaceholder />}
+      />
+    )}
+  />
+ */
+
+export const Tree = <T,>(props: TreeProps<T>) => {
+  const { data, className, nodeRenderer, getNode, ...rest } = props;
+  useTheme();
+
+  const treeRef = React.useRef<HTMLUListElement>(null);
+
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLUListElement>) => {
```

Comment:
so if I started to navigate with keyboard in tree and then press tab, it gets me to the neares next checkbox. But how to get back to the tree navigation? Need to review keyboard..

Reply:
Fixed keyboard to work with checkbox and button

Reply:
I dont know.. experience is weird, now there is no way to quickly go through the nodes.. It hits checkbox and expansion button on every node, when I press arrow keys or tab

Reply:
Left/Right and Up/Down does same. Just Left/Right also expand/collapse.
What did you want to do with this 
![image](https://user-images.githubusercontent.com/81580355/151310214-f2ed7a5c-b61c-4ddf-b580-24260065e647.png)

Reply:
Cant we make it intuitive so it matches the visual layout? Something like this:

- Up/Down will always traverse between nodes
- Left/Right will traverse between focusable elements inside the nodes.
  - Once you reach the leftmost focusable element (e.g. checkbox), pressing Left again will collapse the node
  - Once you reach the rightmost focusable element, pressing Right again will expand the node

W3C does not have recommendations for extra interactable elements inside the nodes, so we have to improvise while still providing a sensible experience.

Reply:
Updated keyboard navigation

Reply:
It is too much work to handle inside node navigation with arrow keys, so I just made focused node inner items focusable with Tab key.